zerver/views/users.py

```diff
@@ -91,17 +92,32 @@ def update_user_backend(request: HttpRequest, user_profile: UserProfile, user_id
     # This condition protected against situations where update_user_backend
     # could cause both is_guest and is_realm_admin to be set.
     # Once we update the frontend to just send a 'role' value, we can remove this check.
-    if (((is_guest is None and target.is_guest) or is_guest) and
-            ((is_admin is None and target.is_realm_admin) or is_admin)):
-        return json_error(_("Guests cannot be organization administrators"))
+    if ((is_guest is None and target.is_guest) or is_guest):
+        if ((is_admin is None and target.is_realm_admin and not target.is_realm_owner) or is_admin):
+            return json_error(_("Guests cannot be organization administrators"))
+
+        if ((is_owner is None and target.is_realm_owner) or is_owner):
+            return json_error(_("Guests cannot be organization owners"))
+
+    if (((is_owner is None and target.is_realm_owner) or is_owner) and
+            ((is_admin is None and target.is_realm_admin and not target.is_realm_owner) or is_admin)):
+        return json_error(_("Organization administrators cannot be organization owners"))
 
     role = None
-    if is_admin is not None and target.is_realm_admin != is_admin:
-        if not is_admin and check_last_admin(user_profile):
-            return json_error(_('Cannot remove the only organization administrator'))
+    if is_owner is not None and target.is_realm_owner != is_owner:
+        if not user_profile.is_realm_owner:
+            return json_error(_("Only organization owner can add or remove owner status"))
+        if not is_owner and check_last_owner(user_profile):
+            return json_error(_("Cannot remove the last organization owner"))
         role = UserProfile.ROLE_MEMBER
+        if is_owner:
+            role = UserProfile.ROLE_REALM_OWNER
+
+    if is_admin is not None and (target.is_realm_admin != is_admin or target.is_realm_owner):
         if is_admin:
             role = UserProfile.ROLE_REALM_ADMINISTRATOR
+        if role is None and not target.is_realm_owner:
+            role = UserProfile.ROLE_MEMBER
 
```

Comment:
Conditions here are bit different and complicated because `is_realm_admin` is true for both admins and owners.

Reply:
I think we should rework how this function works, as the new version is unmanageable.  I'd like us to create a `def parse_requested_role(is_owner, is_admin, is_guest)` that just looks at those properties to determine what role was requested (if any of them is not None) and returns that.  And then we can do a single conditional comparing it with `user_profile.role` to determine whether we need to change the role.

Once we've isolated that, we can open a separate follow-up issue for allowing accepting a `role` parameter in the API, with that `parse_requested_role` function present for backwards-compatibility.

We should do this as a new first commit with just the `is_admin` and `is_guest` parameters, that we can merge happily, and then a small tweak on top of that to pass the new `is_owner` option.

Reply:
Where we should add checks for ensuring that existing role is passed False? (Checking those in this new function will also need target to be passed in it.)